### ISSUE TITLE
Create PQRS measure rules

### ISSUE THREAD

**Issue by leebc** [aethelwulffe/PQRS_Gateway#12]
--==>> Create rules for the PQRS measures and measure groups.

==================================================
From @Lanadot :

**Measure Groups that we have reported for in 2015 :**
* Acute Otitis Externa(AOE);
* Coronary Artery Disease (CAD), 
* Chronic Obstructive Pulmonary Disease (COPD);
* Dementia; 
* Diabetes;
* General Surgery;
* Heart Failure (HF);
* HIV/AIDS;
* Optimizing Patient Exposure to Ionizing Radiation (OPEIR);
* Parkinson’s Disease;
* Rheumatoid Arthritis (RA);
* Sinusitis;
* Total Knee Replacement (TKR).

The specifications defining the eligible patients for the patient sample are on the first page of each measures group overview. 

**Individual Measures that we have reported  for PQRS 2015:**
1, 5, 6, 7, 8, 21, 22, 23, 24, 32, 33, 39, 40, 41, 46, 47, 48, 50, 51, 52, 53, 54, 
76, 91, 93, 109, 110, 111, 112, 113, 116, 117, 118, 119, 121, 122, 126, 127, 
128, 130, 131, 134, 145, 146, 147, 154, 155, 163, 173, 178, 179, 181, 182, 185 
187, 193, 204, 205, 225, 226, 236, 238, 242, 261, 265, 317, 320, 322, 323, 324,
 325, 331, 332, 333, 334, 343, 358, 387, 390,  398, 400, 401

The Individual Measures should take priority.

==================================================

These rules exist here:
[https://github.com/aethelwulffe/PQRS_Gateway/tree/PQRS_incomplete_rules/library/classes/rulesets/PQRS/reports](https://github.com/aethelwulffe/PQRS_Gateway/tree/PQRS_incomplete_rules/library/classes/rulesets/PQRS/reports)

**IMPORTANT NOTE:**  This is in branch **PQRS_incomplete_rules**.  There is also  a branch **PQRS_completed_rules**, which should currently be empty except for a "**PQRS_always_met**" test rule.  The plan is that each rule will be worked on and committed separately in **PQRS_incomplete_rules**; then, when they're finished, these commits will be **git cherrypick**ed  into **PQRS_completed_rules**, and eventually merged into the release.

Each rule consists of a top level **PQRS_0###.php** file, a **PQRS_0###/** directory, and sub-files: **Denominator.php**,  **InitialPatientPopulation.php**,  **PopulationCriteria.php*, **Exclusion.php**, and **Numerator.php**.  There can be multiples of these sub-files if needed.  I.E. in a case where there are multiple Numerators.  
Please be certain that changes to each rule are committed separately to make it easier to cheerypick them into **PQRS_completed_rules**.

See the "Hints" file for hints/tricks/tips on creating rules.  

Please see my follow-up comments for additional information.

**Comment by leebc** [comment]
I have created [https://github.com/aethelwulffe/PQRS_Gateway/tree/PQRS_incomplete_rules/library/classes/rulesets/PQRS/reports/Template](https://github.com/aethelwulffe/PQRS_Gateway/tree/PQRS_incomplete_rules/library/classes/rulesets/PQRS/reports/Template).

Which contains a template for PQRS Measure Rules and a bash script, **mkpqrsskel**, that will create a directory and copy the template into properly named files.  I.e.: **PQRS_0005.php**.  It also edits the files so that the code references the specified measure number.  I.e.:

> class PQRS_0005 extends AbstractPQRSReport


Please refer to the [https://github.com/aethelwulffe/PQRS_Gateway/blob/PQRS_incomplete_rules/library/classes/rulesets/PQRS/reports/Hints](https://github.com/aethelwulffe/PQRS_Gateway/blob/PQRS_incomplete_rules/library/classes/rulesets/PQRS/reports/Hints) file.

Usage (from library/classes/rulesets/PQRS/reports/):
> $ **Template/mkpqrsskel 0005**
MeasureNumber is 0005
DestinationDirectory is PQRS_0005/
Source_FILE: Template/PQRS_template.php Destination_FILE: PQRS_0005.php
Source file: Template/PQRS_template/Denominator.php
Destination file: PQRS_0005//Denominator.php
Source file: Template/PQRS_template/Exclusion.php
Destination file: PQRS_0005//Exclusion.php
Source file: Template/PQRS_template/InitialPatientPopulation.php
Destination file: PQRS_0005//InitialPatientPopulation.php
Source file: Template/PQRS_template/Numerator.php
Destination file: PQRS_0005//Numerator.php
Source file: Template/PQRS_template/PopulationCriteria.php
Destination file: PQRS_0005//PopulationCriteria.php
PQRS_0005/
Denominator.php  InitialPatientPopulation.php  PopulationCriteria.php
Exclusion.php	 Numerator.php

(Don't worry too much about this, I'm going to go ahead and try to create skeletons for all the measure rules tonight.)

**Comment by leebc** [comment]
Please see [https://github.com/aethelwulffe/PQRS_Gateway/issues/10](https://github.com/aethelwulffe/PQRS_Gateway/issues/10) for information regarding 
NFQ_0028_2014.php
NFQ_0028a.php
NFQ_0028b.php
PQRS_0001.php
PQRS_0005.php
PQRS_0066.php
PQRS_0110.php
PQRS_0111.php
PQRS_0128.php

**Comment by meggerc** [comment]
The group measures are composed of individual measures, so the individual measures have to be done first. I'm focusing on the individual measures that are commonly used per Lana.

I've run into some measures that can be reported as part of a group but not individually, e.g, Measure #108, part of the Rheumatoid Arthritis group.

I'm working on a table that includes:
the individual PQRS#, 
the group measure it can be reported with
the domain, e.g., effective or efficient care
and whether the measure can be reported individually or part of a group.
I created numeric codes for the domains and measure group names.
I'll post that soon. 

The individual specification manual table of contents has links that go to the rules for each individual measure. 
[2015_PQRS_IndividualMeasureSpec_ClaimsRegistry_122314.pdf](https://github.com/aethelwulffe/PQRS_Gateway/files/128888/2015_PQRS_IndividualMeasureSpec_ClaimsRegistry_122314.pdf)

**Comment by leebc** [comment]
Used **mkpqrsskel** script to create skeletons for rules Lana said were reported in 2015:

PQRS_0001 PQRS_0005 PQRS_0006 PQRS_0007 PQRS_0008 PQRS_0021 PQRS_0022 PQRS_0023 PQRS_0024 PQRS_0032 PQRS_0033 PQRS_0039 PQRS_0040 PQRS_0041 PQRS_0046 PQRS_0047 PQRS_0048 PQRS_0050 PQRS_0051 PQRS_0052 PQRS_0053 PQRS_0054 PQRS_0066  PQRS_0076 PQRS_0091 PQRS_0093 PQRS_0109 PQRS_0110 PQRS_0111 PQRS_0112 PQRS_0113 PQRS_0116 PQRS_0117 PQRS_0118 PQRS_0119 PQRS_0121 PQRS_0122 PQRS_0126 PQRS_0127 PQRS_0128 PQRS_0130 PQRS_0131 PQRS_0134 PQRS_0145 PQRS_0146 PQRS_0147 PQRS_0154 PQRS_0155 PQRS_0163 PQRS_0173 PQRS_0178 PQRS_0179 PQRS_0181 PQRS_0182 PQRS_0185 PQRS_0187 PQRS_0193 PQRS_0204 PQRS_0205 PQRS_0225 PQRS_0226 PQRS_0236 PQRS_0238 PQRS_0242 PQRS_0261 PQRS_0265 PQRS_0317 PQRS_0320 PQRS_0322 PQRS_0323 PQRS_0324  PQRS_0325 PQRS_0331 PQRS_0332 PQRS_0333 PQRS_0334 PQRS_0343 PQRS_0358 PQRS_0387  PQRS_0390 PQRS_0398 PQRS_0400 PQRS_0401.

This listing also includes rules I had previously copied code for.  (See [https://github.com/aethelwulffe/PQRS_Gateway/issues/10](https://github.com/aethelwulffe/PQRS_Gateway/issues/10))  The existing rules were accidentally overwritten during this automated creation/commit process, but I think I was able to correctly revert those commits.

**Comment by aethelwulffe** [comment]
The README for measure #1 now contains some useful PHP/MYSQL that could be used to complete the measure.
  The actual time I spent writing the queries to return true/false for this fairly complex measure was ~35 minutes.  All that remains is to thump them into an appropriately named function.  My only questions involve the naming of the functions and their incoming arguments.

**Comment by leebc** [comment]
I have created README.TXT s for Measures PQRS_0001, PQRS_0005, PQRS_0006, PQRS_0007, PQRS_0008, PQRS_0021, and PQRS_0022.
The READMEs containa write-up of the Measures, copied from the PDF with as much extraneous information removed as seems plausible.
Further, the write-up attempts to illustrate the logic needed to implement each rule.  This involves identifying each element that must be checked for.

**Comment by aethelwulffe** [comment]
Completed initial draft of rules for pqrs1-8.  21 and 22 are in the 
works, with 21 getting temporarily skipped (some late-night 
miscommunication).  21 and 22 are getting temporary table code (for now) 
since the criteria involve hundreds and hundreds of codes.  It will 
probably be best to move these into permanent tables for our purposes.
   Once again I am having "why" questions about a bit in OpenEMR called 
'sqlStatement'.  I also want to check out the state of PDO and mysqli 
for openemr.  I know there must be some new stuff coming down the pipe 
with PHP7, and I don't want to get deprecated...and I don't want to not 
have things work because we aren't compatible with deprecated functions 
either.
Meg, the README's have been invaluable.

**Comment by meggerc** [comment]
I was working on table code for the denominator codes in table form and
then Bryan requested them in CSV row form, so I switched over.

Do you want both? I have an excel sheet with all but about 20 of the
individual measures as a CSV. I need to finish the 20 and proof all before
I send it.

I also have a reference table that notes the PQRS and NQF# for each
measure, the domain, group measure category (if applicable) and a short
description.

If you agree this is what I'll do: I'll create  code tables for the
individual measures by domain and create a CSV file that Bryan requested.

Once this is complete, I'll do the same for the group measures. There are
some group measures that aren't listed in the individual measure manual.

Which README's are you referring to? Thanks!

**Comment by leebc** [comment]
@meggerc : I've been creating READMEs in each measure directory.   (I've only gotten to #22.)  
Example:
[https://github.com/aethelwulffe/PQRS_Gateway/blob/PQRS_incomplete_rules/library/classes/rulesets/PQRS/reports/PQRS_0022/README.TXT](https://github.com/aethelwulffe/PQRS_Gateway/blob/PQRS_incomplete_rules/library/classes/rulesets/PQRS/reports/PQRS_0022/README.TXT)

**Comment by meggerc** [comment]
I'm looking forward to looking them over after work tonight. :) Thanks!

**Comment by aethelwulffe** [comment]
Meg, since you have gotten through 20 of them, that will go a really long way against getting some more coded this weekend.  Frankly, the part you have been doing takes longer and is harder than writing the actual code!  It will be a breeze with organized data.
  Afterward, we will all have a better idea of what is most needed for completing the process.  Having the tabled data along with an algebraic shorthand of how the measure is calculated gives us another leg up on the code verification, creating target test data, and future revisions.  It is also of HUGE help when a measure has so many codes in it that we have to put them in a temporary database table or even a permanent one.
  Let's head directly into writing code.  Once we have burned up our prepared source material, our efficiency will improve through familiarity.  If we have several of us coding beyond that point, I can shift to writing a unit test that we can debug with and start the cleanup phase on each measure.
Plain old Signature:  Art

**Comment by meggerc** [comment]
Sure I'll do whatever is needed. I can come in anytime tomorrow if you want
to go over stuff with me or we can just do it through github

**Comment by aethelwulffe** [comment]
Oh... ALL but 20!  Not just 20!  Awesome!  Prepping just twenty in a pile of work...all but twenty is a significant portion of the whole project!
  What I get for skimming.
Plain old Signature:  Art

**Comment by meggerc** [comment]
Lol

**Comment by meggerc** [comment]
It isn't that much work so re think it and let me know

**Comment by aethelwulffe** [comment]
I was hoping for an in shop workday.  I would love to get both you and Bclee writing code.  You, because you want to, bclee and Floyd because they can...even if they think they can't.  With all that material prepped, no need not to.
Plain old Signature:  Art

**Comment by meggerc** [comment]
Yeah I'm only free on weekends now I'm free tomorrow if anyone else is

**Comment by aethelwulffe** [comment]
I did a 24 hour sleep-in, so I'll be there in the wee hours for anyone to come in at any time tomorrow.  Gonna do server maintenance, then some cleanup on the code we have already committed.
Plain old Signature:  Art

**Comment by meggerc** [comment]
I'll be sleeping in the wee hours but I can be there around 11.

**Comment by meggerc** [comment]
As of Saturday, Art and I decided to create tables for the ICD, CPT and HCPCS codes instead of including each code in the queries. Some of the measures have over 2,000 codes; this eliminates the need for formatting and results in cleaner code. I've sent a CSV file to Art containing the codes for the Person and Caregiver-Centered Experience and Outcomes (PTCT).

![image](https://cloud.githubusercontent.com/assets/16786878/13203403/65421b34-d885-11e5-832a-ff417be8cbb4.png)

This is how it looks:
![image](https://cloud.githubusercontent.com/assets/16786878/13203468/f7ab6600-d886-11e5-9f72-0371a7de1002.png)

Art will make sure this works out and if so, I'll continue with the rest of the measures and send them over by Domain. 

![image](https://cloud.githubusercontent.com/assets/16786878/13203452/a87b7d0e-d886-11e5-8ff3-cf3fbb53fdeb.png)

**Comment by aethelwulffe** [comment]
Hey Meg!  Where/how did you send the file?  FTP?  Ain't got an email from you, but a lot of folks have been saying the gmail gets bounces (on occasion) sending to us during the last week.

**Comment by aethelwulffe** [comment]
BTW, told Brian about this last night.  This is going to be such a huge game changer...provided that testing you mention works out.  The code will be OH-SO much easier to debug as well.  It isn't really all that hard, but there is just potentially so much of it.

**Comment by meggerc** [comment]
I sent it to your email and star frontiers I just forwarded it again

**Comment by aethelwulffe** [comment]
Didn't get it.  I'll check the server...

**Comment by aethelwulffe** [comment]
Try sending the file in a zip if you would.  We could always use the FTP.

**Comment by aethelwulffe** [comment]
Hi Meg!
   Please give me a call when you can.  I wanna get that file and get to 
work!  I did some testing, and I think we may be ok on this approach.

**Comment by meggerc** [comment]
Art - I'll look for ftp log in information later. For now, here is the first table for the Person centered domain:
[PTCT_Domain_Code_Table.xlsx](https://github.com/aethelwulffe/PQRS_Gateway/files/140797/PTCT_Domain_Code_Table.xlsx)

**Comment by meggerc** [comment]
Art  - I found my user name and password for FileZilla - what do I use for host? thanks! :)

**Comment by aethelwulffe** [comment]
I think you had Fireftp set up in firefox with the host name.
I'll send the host in a direct email.

**Comment by meggerc** [comment]
Here's the text version.
[PTCT_Domain_Code_Table.txt](https://github.com/aethelwulffe/PQRS_Gateway/files/141062/PTCT_Domain_Code_Table.txt)

**Comment by meggerc** [comment]
Here is the code for the Communication and Care Coordination domain:
[CCCO_Domain_Code_Table.txt](https://github.com/aethelwulffe/PQRS_Gateway/files/141496/CCCO_Domain_Code_Table.txt)

**Comment by meggerc** [comment]
Note that CCCO Domain individual measures 219-223 use the same codes for our purposes.

**Comment by leebc** [comment]
Created README.TXT for PQRS_0023 and PQRS_0024.
These have HUGE lists of codes as well. 

PQRS_0024 has 2 sets of criteria, but only 1 numerator.  I THINK this can be built as one Denominator.php/InitialPatientPopulation.php
@aethelwulffe , let me know if you want me to create two Denominator.php and InitialPatientPopulation.php for this measure.

**Comment by meggerc** [comment]
Efficiency and cost reduction domain. Note that pqrs_0322-pqrs_0324 use the same code; and 331-334 use the same code.
[ECR_Domain_Code_Table.txt](https://github.com/aethelwulffe/PQRS_Gateway/files/141557/ECR_Domain_Code_Table.txt)

**Comment by leebc** [comment]
Created README.TXT for PQRS_0032.  Pretty straightforward.

**Comment by meggerc** [comment]
Community/Population Heatlh Domain - these are all pretty straightforward if you need a break from more complicated rules.
[POPH_Domain_Code_Table.txt](https://github.com/aethelwulffe/PQRS_Gateway/files/141568/POPH_Domain_Code_Table.txt)

**Comment by aethelwulffe** [comment]
Trying to turn on computer.  Not having great success.  Swear to get something done in the morning.

**Comment by meggerc** [comment]
Patient Safety Domain - #383 is the messiest measure I've seen so far..
[PTSF_Domain_Code_Table.txt](https://github.com/aethelwulffe/PQRS_Gateway/files/141653/PTSF_Domain_Code_Table.txt)

**Comment by aethelwulffe** [comment]
@meggerc 
Trying to label these...some of the columns are a little confusing in the original file.  I have downloaded the text versions and I will continue piecing them into a two-column table using the pqrs_0001a format.

**Comment by meggerc** [comment]
I can fix them - which file are you referring to? and how can I make it
less confusing?

**Comment by aethelwulffe** [comment]
Well, we have a couple of little issues.
First, the trailing zeros after the decimal places are being dropped by 
formatting (apparently).
The naming convention would probably be most helpful in something like this:
Where the expression is:
For measure 100:
((One code list 'a' AND one code list 'b')
                             OR
(One code list 'a' AND (one code list 'c' + modifier 'cm'))
AND
one code list d [a CPTII code] WITHOUT modifier 'dm'

Which is ((a+b) OR (a+(c+x))) + (d NOT+dm)

We could then use the following labels

pqrs_0100a
pqrs_0100b
pqrs_0100c
pqrs_0100cm
pqrs_0100d
pqrs_0100dm

Any types of codes can be mixed in a list (ICD9/10/CPT etc) so long as 
you only need any matching code.

**Comment by meggerc** [comment]
I can redo the tables I've sent. I'm working on an excel formula to fix the
trailing zero issue.

Do you want to start over once I create the tables with new labels and the
trailing zeroes?

**Comment by meggerc** [comment]
Art - I'm going to call you at 115 - try to be available please. :)

**Comment by aethelwulffe** [comment]
You got it meg.

**Comment by meggerc** [comment]
@aethelwulffe I just found this file in the getting started section on the CMS QPRS website. It contains a column for the measure # and a column for the code. Do you think you can use this create the tables.  
[2015_PQRS_IndivMeasures_SingleSource_122214_v9.1.xlsx](https://github.com/aethelwulffe/PQRS_Gateway/files/143480/2015_PQRS_IndivMeasures_SingleSource_122214_v9.1.xlsx)

**Comment by meggerc** [comment]
@aethelwulffe I'm redoing the tables with the row labels you requested in CSV form. Ignore the tables that are already posted in this thread. I'll proof the new files for trailing zeros but I don't think it's a problem.

**Comment by aethelwulffe** [comment]
Incredibly awesome starting point!

**Comment by aethelwulffe** [comment]
That single column file is exactly what we need ain't it?  Just gotta name them right, and we are gold.  Awesome find.
  I am taking cold medicine, whiskey, or anything else I got to so I get something done tomorrow.  Bryan may be down for the count though.  He was doing extremely poorly this morning.

**Comment by FloydERoberts3** [comment]
This looks really great - conceptually queries will step through the tables
registering hits? If this is correct will the hits registered be in a table
a straight string or some strange key value pair arrangement?
really progressive looking stuff!

**Comment by aethelwulffe** [comment]
Key value pair essentially
Basic vertical table setup.  We are busting them up into smaller table by reporting 'domain', so we are dealing with fewer rows.  It may be worth looking at to bust the tables up by 'a', 'b', 'c'.

**Comment by FloydERoberts3** [comment]
Small question - Data from lou Has repeated variety of PQRS "Measure
Number" Spec indicates that repeated array is the measure details - which
is repeated for for that measure number.

Question: are their multiple measure numbers reported for a single measure
group? if so are there multiple measure details for a given measure number
?
Making progress

**Comment by meggerc** [comment]
@FloydERoberts3 Measure groups include individual measures. There are manual specs for both individual measure reporting and group measure reporting. I don't understand this question, can you reword it?... if so are there multiple measure details for a given measure number?

**Comment by FloydERoberts3** [comment]
Great yes
Question:
are there multiple measure details for a given measure number?

**Comment by meggerc** [comment]
Oh, like and/or types of things? Yes.

**Comment by meggerc** [comment]
They're all in the manual and some Read Me's have been created.

**Comment by aethelwulffe** [comment]
Yes to all above

**Comment by aethelwulffe** [comment]
On measure 146.
44 initial commits done, 39 to go for the "priority" sets.
Should get done tomorrow, then on to adding the code tables day after tomorrow.  Hopefully...possibly...create some patient encounters and test the raw queries against the tables to see just how stupid the code is.
UI, direct entry, and xml export after that.  Luckily, the numerator section of our measures provides EXACTLY what we need to use to satisfy the direct entry values.

**Comment by aethelwulffe** [comment]
So...
tweaked the tables a bit during testing, got a couple loaded into a local test instance, then checked a few queries from the measure code against them....
# THE CODE WORKS!!!!!
Any refactoring of the measures will be for optimization or plain old bug/typo issues (which could possibly happen you know) and not because the logic is whacko.  I was assuming the logic was whacko.
It isn't.
Successful test completed 4.5 hours before March 1st.
